File: clients/android/app/src/main/kotlin/com/lextures/android/core/search/SearchPathNavigator.kt
package com.lextures.android.core.search

import com.lextures.android.core.navigation.MoreDestination
import com.lextures.android.core.navigation.ShellTab
import com.lextures.android.core.routing.DeepLinkDestination
import com.lextures.android.core.routing.DeepLinkRouter

sealed class SearchNavigationTarget {
    data class ShellTabTarget(val tab: ShellTab) : SearchNavigationTarget()
    data class DeepLinkTarget(val destination: DeepLinkDestination) : SearchNavigationTarget()
    data class MoreTarget(val destination: MoreDestination) : SearchNavigationTarget()
}

object SearchPathNavigator {
    fun resolve(rawPath: String): SearchNavigationTarget? {
        val trimmed = rawPath.trim()
        if (trimmed.isEmpty()) return null

        if (trimmed.startsWith("lextures://shell/")) {
            return when (trimmed.removePrefix("lextures://shell/").lowercase()) {
                "teach" -> SearchNavigationTarget.ShellTabTarget(ShellTab.Teach)
                "children" -> SearchNavigationTarget.ShellTabTarget(ShellTab.Children)
                "calendar" -> SearchNavigationTarget.ShellTabTarget(ShellTab.Calendar)
                else -> null
            }
        }

        val path = trimmed.substringBefore('?')
        val segments = path.trim('/').split('/').filter { it.isNotEmpty() }

        if (segments.isEmpty()) {
            return SearchNavigationTarget.ShellTabTarget(ShellTab.Home)
        }

        return when (segments.first().lowercase()) {
            "courses" -> {
                if (segments.size == 1) {
                    SearchNavigationTarget.ShellTabTarget(ShellTab.Courses)
                } else {
                    SearchNavigationTarget.DeepLinkTarget(DeepLinkRouter.resolve(path))
                }
            }
            "inbox" -> SearchNavigationTarget.DeepLinkTarget(DeepLinkDestination.Inbox)
            "notebooks" -> SearchNavigationTarget.ShellTabTarget(ShellTab.Notebooks)
            "calendar" -> SearchNavigationTarget.ShellTabTarget(ShellTab.Calendar)
            "parent" -> SearchNavigationTarget.ShellTabTarget(ShellTab.Children)
            "todos" -> SearchNavigationTarget.MoreTarget(MoreDestination.Planner)
            "portfolios" -> SearchNavigationTarget.MoreTarget(MoreDestination.Portfolio)
            "catalog" -> SearchNavigationTarget.MoreTarget(MoreDestination.Catalog)
            "paths" -> SearchNavigationTarget.MoreTarget(MoreDestination.Paths)
            "library" -> SearchNavigationTarget.MoreTarget(MoreDestination.Library)
            "reading" -> SearchNavigationTarget.MoreTarget(MoreDestination.Reading)
            "review" -> SearchNavigationTarget.DeepLinkTarget(DeepLinkDestination.Review)
            "credentials" -> SearchNavigationTarget.MoreTarget(MoreDestination.Credentials)
            "advising" -> SearchNavigationTarget.MoreTarget(MoreDestination.Advising)
            "peer-reviews" -> SearchNavigationTarget.MoreTarget(MoreDestination.PeerReviews)
            "report-cards" -> SearchNavigationTarget.MoreTarget(MoreDestination.ReportCards)
            "settings" -> SearchNavigationTarget.ShellTabTarget(ShellTab.Profile)
            else -> {
                val deep = DeepLinkRouter.resolve(path)
                if (deep is DeepLinkDestination.Home && segments.first().lowercase() != "courses") {
                    null
                } else {
                    SearchNavigationTarget.DeepLinkTarget(deep)
                }
            }
        }
    }
}

File: clients/ios/Lextures/Core/Search/SearchPathNavigator.swift
import Foundation

enum SearchNavigationTarget: Equatable {
    case shellTab(ShellTab)
    case deepLink(DeepLinkDestination)
    case more(MoreDestination)
}

/// Maps web-style search result paths to native shell navigation.
enum SearchPathNavigator {
    static func resolve(_ rawPath: String) -> SearchNavigationTarget? {
        let trimmed = rawPath.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        if trimmed.hasPrefix("lextures://shell/") {
            let segment = String(trimmed.dropFirst("lextures://shell/".count)).lowercased()
            switch segment {
            case "teach": return .shellTab(.teach)
            case "children": return .shellTab(.children)
            case "calendar": return .shellTab(.calendar)
            default: break
            }
        }

        let path = trimmed.split(separator: "?").first.map(String.init) ?? trimmed
        let segments = path.split(separator: "/").map(String.init).filter { !$0.isEmpty }

        if segments.isEmpty || (segments.count == 1 && segments[0].isEmpty) {
            return .shellTab(.home)
        }

        switch segments[0].lowercased() {
        case "courses":
            if segments.count == 1 { return .shellTab(.courses) }
            return .deepLink(DeepLinkRouter.resolve(path))
        case "inbox":
            return .deepLink(.inbox)
        case "notebooks":
            return .shellTab(.notebooks)
        case "calendar":
            return .shellTab(.calendar)
        case "parent":
            return .shellTab(.children)
        case "todos":
            return .more(.planner)
        case "portfolios":
            return .more(.portfolio)
        case "catalog":
            return .more(.catalog)
        case "paths":
            return .more(.paths)
        case "library":
            return .more(.library)
        case "reading":
            return .more(.reading)
        case "credentials":
            return .more(.credentials)
        case "advising":
            return .more(.advising)
        case "settings":
            return .shellTab(.profile)
        default:
            let deep = DeepLinkRouter.resolve(path)
            if case .home = deep, segments.first?.lowercased() != "courses" {
                return nil
            }
            return .deepLink(deep)
        }
    }
}